File: src/game/items.ts
// === Items & Inventory System ===

export const enum ItemType {
	Weapon = 0,
	Armor = 1,
	Potion = 2,
	Food = 3,
	Material = 4,
	Misc = 5,
}

export type Item = {
	id: string;
	name: string;
	type: ItemType;
	value: number; // Gold value
	quantity: number;
	icon: string; // Emoji icon for display
	description: string;
	// Optional stat bonuses when equipped/consumed
	effect?: {
		hp?: number;
		mp?: number;
		sp?: number;
		str?: number;
		end?: number;
		agi?: number;
	};
};

export type Inventory = {
	items: Item[];
	maxSlots: number; // Grid capacity
};

export function createInventory(maxSlots = 64): Inventory {
	return {items: [], maxSlots};
}

export function addItem(inv: Inventory, item: Item): boolean {
	// Try stacking with existing
	const existing = inv.items.find(i => i.id === item.id);
	if (existing) {
		existing.quantity += item.quantity;
		return true;
	}

	if (inv.items.length >= inv.maxSlots) {
		return false; // Inventory full
	}

	inv.items.push({...item});
	return true;
}

export function removeItem(inv: Inventory, itemId: string, quantity = 1): boolean {
	const idx = inv.items.findIndex(i => i.id === itemId);
	if (idx === -1) {
		return false;
	}

	const item = inv.items[idx];
	if (item.quantity < quantity) {
		return false;
	}

	item.quantity -= quantity;
	if (item.quantity <= 0) {
		inv.items.splice(idx, 1);
	}

	return true;
}

export function hasItem(inv: Inventory, itemId: string): boolean {
	return inv.items.some(i => i.id === itemId && i.quantity > 0);
}

export function itemCount(inv: Inventory): number {
	return inv.items.reduce((sum, i) => sum + i.quantity, 0);
}

// === Item Database ===

/**
 * Item blueprint — static definition. Quantity is set at creation time.
 * To add a new item: add one entry to ITEM_CATALOG. Everything else adapts.
 */
export type ItemDef = Omit<Item, 'quantity'>;

/** Central item catalog — single source of truth for all item definitions. */
export const ITEM_CATALOG: Record<string, ItemDef> = {
	// ── Currency / Resources ──
	gold: {
		id: 'gold', name: 'Gold', type: ItemType.Misc, value: 1,
		icon: '\u{1FA99}', description: 'Universal currency',
	},

	// ── Consumables ──
	potion_hp: {
		id: 'potion_hp', name: 'Health Potion', type: ItemType.Potion, value: 50,
		icon: '\u2764', description: 'Restores 30 HP', effect: {hp: 30},
	},
	potion_mp: {
		id: 'potion_mp', name: 'Mana Potion', type: ItemType.Potion, value: 75,
		icon: '\u2728', description: 'Restores 15 MP', effect: {mp: 15},
	},
	food_bread: {
		id: 'food_bread', name: 'Bread', type: ItemType.Food, value: 10,
		icon: '\u{1F35E}', description: 'Restores 10 HP', effect: {hp: 10},
	},

	// ── Materials ──
	mat_wood: {
		id: 'mat_wood', name: 'Wood', type: ItemType.Material, value: 5,
		icon: '\u{1FAB5}', description: 'Building material',
	},
	mat_iron: {
		id: 'mat_iron', name: 'Iron Ore', type: ItemType.Material, value: 15,
		icon: '\u26CF', description: 'Smithing material',
	},
	mat_herb: {
		id: 'mat_herb', name: 'Herb', type: ItemType.Material, value: 8,
		icon: '\u{1F33F}', description: 'Alchemy ingredient',
	},

	// ── Equipment ──
	wpn_dagger: {
		id: 'wpn_dagger', name: 'Rusty Dagger', type: ItemType.Weapon, value: 30,
		icon: '\u{1F5E1}', description: '+2 STR when equipped', effect: {str: 2},
	},
	arm_leather: {
		id: 'arm_leather', name: 'Leather Armor', type: ItemType.Armor, value: 60,
		icon: '\u{1F6E1}', description: '+2 END when equipped', effect: {end: 2},
	},

	// ── Valuables ──
	misc_gem: {
		id: 'misc_gem', name: 'Gemstone', type: ItemType.Misc, value: 100,
		icon: '\u{1F48E}', description: 'Valuable gem, can be sold',
	},
};

/** Create an item instance from the catalog by ID. */
export function makeItem(id: string, qty = 1): Item {
	const def = ITEM_CATALOG[id];
	if (!def) {
		return {
			id, name: id, type: ItemType.Misc, value: 0, quantity: qty,
			icon: '?', description: 'Unknown item',
		};
	}

	return {...def, quantity: qty};
}

// ── Legacy factory wrappers (backward-compatible, delegate to catalog) ──

export function makePotion(qty = 1): Item {
	return makeItem('potion_hp', qty);
}

export function makeMpPotion(qty = 1): Item {
	return makeItem('potion_mp', qty);
}

export function makeBread(qty = 1): Item {
	return makeItem('food_bread', qty);
}

export function makeWood(qty = 1): Item {
	return makeItem('mat_wood', qty);
}

export function makeIronOre(qty = 1): Item {
	return makeItem('mat_iron', qty);
}

export function makeRustyDagger(): Item {
	return makeItem('wpn_dagger');
}

export function makeLeatherArmor(): Item {
	return makeItem('arm_leather');
}

export function makeHerb(qty = 1): Item {
	return makeItem('mat_herb', qty);
}

export function makeGem(qty = 1): Item {
	return makeItem('misc_gem', qty);
}

/**
 * Loot table entry: item catalog key, chance (0–1), and quantity range.
 * Used by generateNpcInventory and generateSettlementInventory.
 */
type LootEntry = {
	item: string;
	chance: number;
	min: number;
	max: number;
	/** Minimum NPC level required (default 0). */
	minLevel?: number;
};

/** Loot tables keyed by NPC type number. */
const NPC_LOOT: Record<number, LootEntry[]> = {
	// Peasant
	0: [
		{
			item: 'food_bread', chance: 0.6, min: 1, max: 3,
		},
		{
			item: 'mat_wood', chance: 0.4, min: 1, max: 4,
		},
		{
			item: 'mat_herb', chance: 0.2, min: 1, max: 2,
		},
	],
	// Woodcutter
	1: [
		{
			item: 'mat_wood', chance: 1, min: 2, max: 7,
		},
		{
			item: 'food_bread', chance: 0.5, min: 1, max: 2,
		},
	],
	// Merchant
	2: [
		{
			item: 'potion_hp', chance: 0.7, min: 1, max: 3,
		},
		{
			item: 'food_bread', chance: 0.6, min: 2, max: 6,
		},
		{
			item: 'potion_mp', chance: 0.5, min: 1, max: 2,
		},
		{
			item: 'mat_iron', chance: 0.4, min: 1, max: 3,
		},
		{
			item: 'misc_gem', chance: 0.3, min: 1, max: 1,
		},
		{
			item: 'wpn_dagger', chance: 0.2, min: 1, max: 1,
		},
	],
	// Caravan
	3: [
		{
			item: 'food_bread', chance: 1, min: 3, max: 7,
		},
		{
			item: 'potion_hp', chance: 0.7, min: 1, max: 3,
		},
		{
			item: 'mat_iron', chance: 0.6, min: 2, max: 5,
		},
		{
			item: 'misc_gem', chance: 0.4, min: 1, max: 2,
		},
	],
	// Bandit
	4: [
		{
			item: 'potion_hp', chance: 0.7, min: 1, max: 2,
		},
		{
			item: 'wpn_dagger', chance: 0.5, min: 1, max: 1, minLevel: 3,
		},
		{
			item: 'misc_gem', chance: 0.4, min: 1, max: 2,
		},
	],
	// Guard
	5: [
		{
			item: 'food_bread', chance: 0.6, min: 1, max: 3,
		},
		{
			item: 'potion_hp', chance: 0.5, min: 1, max: 1,
		},
		{
			item: 'arm_leather', chance: 0.3, min: 1, max: 1, minLevel: 3,
		},
	],
	// Witch
	6: [
		{
			item: 'potion_mp', chance: 1, min: 1, max: 3,
		},
		{
			item: 'mat_herb', chance: 0.7, min: 2, max: 5,
		},
		{
			item: 'potion_hp', chance: 0.5, min: 1, max: 2,
		},
	],
	// Sorceress
	7: [
		{
			item: 'potion_mp', chance: 1, min: 2, max: 5,
		},
		{
			item: 'potion_hp', chance: 1, min: 1, max: 3,
		},
		{
			item: 'mat_herb', chance: 0.6, min: 3, max: 7,
		},
		{
			item: 'misc_gem', chance: 0.4, min: 1, max: 2,
		},
	],
};

function rollLoot(entries: LootEntry[], level: number, rng: () => number): Item[] {
	const items: Item[] = [];
	for (const entry of entries) {
		if (entry.minLevel && level < entry.minLevel) {
			continue;
		}

		if (rng() < entry.chance) {
			const qty = entry.min + Math.floor(rng() * (entry.max - entry.min + 1));
			items.push(makeItem(entry.item, qty));
		}
	}

	return items;
}

// Generate random loot for an NPC based on type and level
export function generateNpcInventory(npcType: number, npcLevel: number, rng: () => number): Item[] {
	const table = NPC_LOOT[npcType];
	if (!table) {
		return [];
	}

	return rollLoot(table, npcLevel, rng);
}

// ── Settlement loot tables (keyed by economy type) ──

const SETTLEMENT_BASE_LOOT: LootEntry[] = [
	{
		item: 'food_bread', chance: 1, min: 5, max: 14,
	},
	{
		item: 'potion_hp', chance: 1, min: 3, max: 9,
	},
];

const SETTLEMENT_ECONOMY_LOOT: Record<string, LootEntry[]> = {
	farming: [
		{
			item: 'food_bread', chance: 1, min: 10, max: 24,
		},
		{
			item: 'mat_herb', chance: 1, min: 5, max: 12,
		},
	],
	mining: [
		{
			item: 'mat_iron', chance: 1, min: 5, max: 14,
		},
		{
			item: 'misc_gem', chance: 1, min: 0, max: 2,
		},
	],
	trade: [
		{
			item: 'potion_hp', chance: 1, min: 5, max: 14,
		},
		{
			item: 'potion_mp', chance: 1, min: 3, max: 9,
		},
		{
			item: 'mat_iron', chance: 1, min: 3, max: 7,
		},
		{
			item: 'misc_gem', chance: 1, min: 0, max: 3,
		},
	],
	fishing: [
		{
			item: 'food_bread', chance: 1, min: 8, max: 19,
		},
		{
			item: 'mat_herb', chance: 1, min: 3, max: 7,
		},
	],
	crafting: [
		{
			item: 'mat_wood', chance: 1, min: 5, max: 14,
		},
		{
			item: 'mat_iron', chance: 1, min: 4, max: 11,
		},
	],
};

// Generate settlement inventory based on population and economy
export function generateSettlementInventory(population: number, economy: string, rng: () => number): Inventory {
	const inv = createInventory(); // Universal 64 slots

	// Base items
	for (const entry of SETTLEMENT_BASE_LOOT) {
		const qty = entry.min + Math.floor(rng() * (entry.max - entry.min + 1));
		if (qty > 0) {
			addItem(inv, makeItem(entry.item, qty));
		}
	}

	// Economy-based items
	const econLoot = SETTLEMENT_ECONOMY_LOOT[economy];
	if (econLoot) {
		for (const entry of econLoot) {
			const qty = entry.min + Math.floor(rng() * (entry.max - entry.min + 1));
			if (qty > 0) {
				addItem(inv, makeItem(entry.item, qty));
			}
		}
	}

	// Population-based bonus items
	const popTier = Math.floor(population / 200);
	if (popTier >= 1) {
		addItem(inv, makeItem('potion_mp', 1 + Math.floor(rng() * popTier)));
	}

	if (popTier >= 2) {
		addItem(inv, makeItem('misc_gem', Math.floor(rng() * popTier)));
	}

	return inv;
}

// Use a consumable item (potion/food) on the player
export function useItem(inv: Inventory, itemId: string, playerCombat: {currentHp: number; maxHp: number; currentMp: number; maxMp: number; currentSp: number; maxSp: number}): string | undefined {
	const item = inv.items.find(i => i.id === itemId);
	if (!item || item.quantity <= 0) {
		return undefined;
	}

	if (item.type !== ItemType.Potion && item.type !== ItemType.Food) {
		return undefined;
	}

	const messages: string[] = [];
	if (item.effect?.hp) {
		const healed = Math.min(item.effect.hp, playerCombat.maxHp - playerCombat.currentHp);
		playerCombat.currentHp += healed;
		messages.push(`+${healed} HP`);
	}

	if (item.effect?.mp) {
		const restored = Math.min(item.effect.mp, playerCombat.maxMp - playerCombat.currentMp);
		playerCombat.currentMp += restored;
		messages.push(`+${restored} MP`);
	}

	if (item.effect?.sp) {
		const restored = Math.min(item.effect.sp, playerCombat.maxSp - playerCombat.currentSp);
		playerCombat.currentSp += restored;
		messages.push(`+${restored} SP`);
	}

	removeItem(inv, itemId, 1);
	return messages.length > 0 ? `Used ${item.name}: ${messages.join(', ')}` : `Used ${item.name}`;
}

File: timaert_c/src/macro/items.cpp
// Faithful port of `src/game/items.ts` (catalog + loot tables + helpers).
//
// All numeric constants, ids, chances, min/max ranges, and gold formulas
// are copied verbatim from the TS source. Adding/removing entries here
// must mirror the TS file 1:1.

#include "macro/items.h"

#include <algorithm>
#include <cmath>
#include <cstring>
#include <string>

namespace sm {

namespace {

// ── Item catalog ───────────────────────────────────────────────
// Order copied verbatim from TS `ITEM_CATALOG`.

constexpr ItemDef kCatalog[] = {
    // Currency / Resources
    {"gold",        "Gold",            ItemType::Misc,        1, 0.01f, "\xF0\x9F\xAA\x99",
        "Universal currency", {}},

    // Consumables
    {"potion_hp",   "Health Potion",   ItemType::Potion,     50, 0.30f, "\xE2\x9D\xA4",
        "Restores 30 HP", {30, 0, 0, 0, 0, 0}},
    {"potion_mp",   "Mana Potion",     ItemType::Potion,     75, 0.30f, "\xE2\x9C\xA8",
        "Restores 15 MP", { 0,15, 0, 0, 0, 0}},
    {"food_bread",  "Bread",           ItemType::Food,       10, 0.10f, "\xF0\x9F\x8D\x9E",
        "Restores 10 HP", {10, 0, 0, 0, 0, 0}},
    {"food_meat",   "Raw Meat",        ItemType::Food,       15, 0.50f, "\xF0\x9F\x8D\x96",
        "Restores 15 HP", {15, 0, 0, 0, 0, 0}},

    // Materials
    {"mat_wood",    "Wood",            ItemType::Material,    5, 2.00f, "\xF0\x9F\xAA\xB5",
        "Building material", {}},
    {"mat_iron",    "Iron Ore",        ItemType::Material,   15, 3.00f, "\xE2\x9B\x8F",
        "Smithing material", {}},
    {"mat_bone",    "Bone",            ItemType::Material,    6, 0.50f, "\xF0\x9F\xA6\xB4",
        "Crafting material from monsters", {}},
    {"mat_hide",    "Hide",            ItemType::Material,   12, 1.00f, "\xF0\x9F\xA7\xB5",
        "Tanned animal hide", {}},
    {"mat_herb",    "Herb",            ItemType::Material,    8, 0.05f, "\xF0\x9F\x8C\xBF",
        "Alchemy ingredient", {}},

    // Equipment
    {"wpn_dagger",  "Rusty Dagger",    ItemType::Weapon,     30, 1.00f, "\xF0\x9F\x97\xA1",
        "+2 STR when equipped", {0, 0, 0, 2, 0, 0}},
    {"arm_leather", "Leather Armor",   ItemType::Armor,      60, 5.00f, "\xF0\x9F\x9B\xA1",
        "+2 END when equipped", {0, 0, 0, 0, 2, 0}},

    // Valuables
    {"misc_gem",    "Gemstone",        ItemType::Misc,      100, 0.05f, "\xF0\x9F\x92\x8E",
        "Valuable gem, can be sold", {}},
};

const std::unordered_map<std::string, const ItemDef*>& catalog_map() {
    static const std::unordered_map<std::string, const ItemDef*> m = []{
        std::unordered_map<std::string, const ItemDef*> r;
        for (const auto& d : kCatalog) r.emplace(d.id, &d);
        return r;
    }();
    return m;
}

// ── Loot tables (copied verbatim from TS) ──────────────────────

struct LootEntry {
    const char* item;
    float       chance;
    int         min;
    int         max;
    int         minLevel;  // 0 = unrestricted
};

// NPC_LOOT[npcType] — keys 0-7 mirror NPCType enum in `npc.h`:
// 0 Peasant, 1 Woodcutter, 2 Merchant, 3 Caravan, 4 Bandit, 5 Guard,
// 6 Witch, 7 Sorceress.

constexpr LootEntry kPeasantLoot[] = {
    {"food_bread", 0.6f, 1, 3, 0},
    {"mat_wood",   0.4f, 1, 4, 0},
    {"mat_herb",   0.2f, 1, 2, 0},
};
constexpr LootEntry kWoodcutterLoot[] = {
    {"mat_wood",   1.0f, 2, 7, 0},
    {"food_bread", 0.5f, 1, 2, 0},
};
constexpr LootEntry kMerchantLoot[] = {
    {"potion_hp",  0.7f, 1, 3, 0},
    {"food_bread", 0.6f, 2, 6, 0},
    {"potion_mp",  0.5f, 1, 2, 0},
    {"mat_iron",   0.4f, 1, 3, 0},
    {"misc_gem",   0.3f, 1, 1, 0},
    {"wpn_dagger", 0.2f, 1, 1, 0},
};
constexpr LootEntry kCaravanLoot[] = {
    {"food_bread", 1.0f, 3, 7, 0},
    {"potion_hp",  0.7f, 1, 3, 0},
    {"mat_iron",   0.6f, 2, 5, 0},
    {"misc_gem",   0.4f, 1, 2, 0},
};
constexpr LootEntry kBanditLoot[] = {
    {"potion_hp",  0.7f, 1, 2, 0},
    {"wpn_dagger", 0.5f, 1, 1, 3},
    {"misc_gem",   0.4f, 1, 2, 0},
};
constexpr LootEntry kGuardLoot[] = {
    {"food_bread",  0.6f, 1, 3, 0},
    {"potion_hp",   0.5f, 1, 1, 0},
    {"arm_leather", 0.3f, 1, 1, 3},
};
constexpr LootEntry kWitchLoot[] = {
    {"potion_mp",  1.0f, 1, 3, 0},
    {"mat_herb",   0.7f, 2, 5, 0},
    {"potion_hp",  0.5f, 1, 2, 0},
};
constexpr LootEntry kSorceressLoot[] = {
    {"potion_mp",  1.0f, 2, 5, 0},
    {"potion_hp",  1.0f, 1, 3, 0},
    {"mat_herb",   0.6f, 3, 7, 0},
    {"misc_gem",   0.4f, 1, 2, 0},
};

struct LootTable {
    const LootEntry* data;
    std::size_t      n;
};

// Index = NPCType integer.
constexpr LootTable kNpcLoot[] = {
    {kPeasantLoot,    sizeof(kPeasantLoot)    / sizeof(LootEntry)},
    {kWoodcutterLoot, sizeof(kWoodcutterLoot) / sizeof(LootEntry)},
    {kMerchantLoot,   sizeof(kMerchantLoot)   / sizeof(LootEntry)},
    {kCaravanLoot,    sizeof(kCaravanLoot)    / sizeof(LootEntry)},
    {kBanditLoot,     sizeof(kBanditLoot)     / sizeof(LootEntry)},
    {kGuardLoot,      sizeof(kGuardLoot)      / sizeof(LootEntry)},
    {kWitchLoot,      sizeof(kWitchLoot)      / sizeof(LootEntry)},
    {kSorceressLoot,  sizeof(kSorceressLoot)  / sizeof(LootEntry)},
};

// FAUNA_LOOT — keyed by faction id string.
constexpr LootEntry kWildlifeLoot[] = {
    {"food_meat", 0.85f, 1, 3, 0},
    {"mat_hide",  0.50f, 1, 2, 0},
};
constexpr LootEntry kDemonsLoot[] = {
    {"mat_bone",  0.70f, 1, 3, 0},
    {"mat_herb",  0.30f, 1, 2, 0},
    {"misc_gem",  0.15f, 1, 1, 3},
};

// Settlement loot.
constexpr LootEntry kSettlementBase[] = {
    {"food_bread", 1.0f,  5, 14, 0},
    {"potion_hp",  1.0f,  3,  9, 0},
};
constexpr LootEntry kEconFarming[] = {
    {"food_bread", 1.0f, 10, 24, 0},
    {"mat_herb",   1.0f,  5, 12, 0},
};
constexpr LootEntry kEconMining[] = {
    {"mat_iron",   1.0f, 5, 14, 0},
    {"misc_gem",   1.0f, 0,  2, 0},
};
constexpr LootEntry kEconTrade[] = {
    {"potion_hp",  1.0f, 5, 14, 0},
    {"potion_mp",  1.0f, 3,  9, 0},
    {"mat_iron",   1.0f, 3,  7, 0},
    {"misc_gem",   1.0f, 0,  3, 0},
};
constexpr LootEntry kEconFishing[] = {
    {"food_bread", 1.0f, 8, 19, 0},
    {"mat_herb",   1.0f, 3,  7, 0},
};
constexpr LootEntry kEconCrafting[] = {
    {"mat_wood",   1.0f, 5, 14, 0},
    {"mat_iron",   1.0f, 4, 11, 0},
};

inline std::vector<ItemStack> roll_loot(const LootEntry* table, std::size_t n,
                                        int level, RngFn rng) {
    std::vector<ItemStack> out;
    out.reserve(n);
    for (std::size_t i = 0; i < n; ++i) {
        const LootEntry& e = table[i];
        if (e.minLevel > 0 && level < e.minLevel) continue;
        if (rng() < e.chance) {
            const int qty = e.min + static_cast<int>(rng() *
                static_cast<float>(e.max - e.min + 1));
            out.push_back({std::string(e.item), qty});
        }
    }
    return out;
}

inline float gold_faction_mult(const char* factionId) {
    if (std::strcmp(factionId, "wildlife") == 0) return 0.1f;
    if (std::strcmp(factionId, "demons")   == 0) return 0.6f;
    if (std::strcmp(factionId, "bandits")  == 0) return 0.8f;
    return 1.0f;
}

} // namespace

// ── Public API ────────────────────────────────────────────────

const ItemDef* item_def(const std::string& id) noexcept {
    const auto& m = catalog_map();
    const auto it = m.find(id);
    return it == m.end() ? nullptr : it->second;
}

float inventory_weight(const Inventory& inv) noexcept {
    float total = 0.0f;
    for (const auto& s : inv.stacks) {
        if (const ItemDef* d = item_def(s.id)) {
            total += d->weight * static_cast<float>(s.count);
        }
    }
    return total;
}

std::vector<ItemStack> generate_npc_inventory(int npcType, int npcLevel, RngFn rng) {
    if (npcType < 0 ||
        npcType >= static_cast<int>(sizeof(kNpcLoot) / sizeof(LootTable))) {
        return {};
    }
    const LootTable& t = kNpcLoot[npcType];
    return roll_loot(t.data, t.n, npcLevel, rng);
}

std::vector<ItemStack> generate_fauna_loot(const char* factionId, int level, RngFn rng) {
    if (std::strcmp(factionId, "wildlife") == 0)
        return roll_loot(kWildlifeLoot, sizeof(kWildlifeLoot) / sizeof(LootEntry), level, rng);
    if (std::strcmp(factionId, "demons") == 0)
        return roll_loot(kDemonsLoot,   sizeof(kDemonsLoot)   / sizeof(LootEntry), level, rng);
    return {};
}

int generate_loot_gold(int level, const char* factionId, RngFn rng) {
    const float mult   = gold_faction_mult(factionId);
    const float base   = (3.0f + static_cast<float>(level) * 2.0f) * mult;
    const float jitter = rng() * base;
    const int v = static_cast<int>(std::floor(base + jitter));
    return v < 0 ? 0 : v;
}

Inventory generate_settlement_inventory(int population, const char* economy, RngFn rng) {
    Inventory inv;

    // Base
    for (const auto& e : kSettlementBase) {
        const int qty = e.min + static_cast<int>(rng() *
            static_cast<float>(e.max - e.min + 1));
        if (qty > 0) inv.add(e.item, qty);
    }

    // Economy-specific
    const LootEntry* econ = nullptr;
    std::size_t      econN = 0;
    if      (std::strcmp(economy, "farming") == 0)  { econ = kEconFarming;  econN = sizeof(kEconFarming) /sizeof(LootEntry); }
    else if (std::strcmp(economy, "mining") == 0)   { econ = kEconMining;   econN = sizeof(kEconMining)  /sizeof(LootEntry); }
    else if (std::strcmp(economy, "trade") == 0)    { econ = kEconTrade;    econN = sizeof(kEconTrade)   /sizeof(LootEntry); }
    else if (std::strcmp(economy, "fishing") == 0)  { econ = kEconFishing;  econN = sizeof(kEconFishing) /sizeof(LootEntry); }
    else if (std::strcmp(economy, "crafting") == 0) { econ = kEconCrafting; econN = sizeof(kEconCrafting)/sizeof(LootEntry); }

    if (econ) {
        for (std::size_t i = 0; i < econN; ++i) {
            const auto& e = econ[i];
            const int qty = e.min + static_cast<int>(rng() *
                static_cast<float>(e.max - e.min + 1));
            if (qty > 0) inv.add(e.item, qty);
        }
    }

    // Population bonus
    const int popTier = population / 200;
    if (popTier >= 1) inv.add("potion_mp", 1 + static_cast<int>(rng() * static_cast<float>(popTier)));
    if (popTier >= 2) inv.add("misc_gem",      static_cast<int>(rng() * static_cast<float>(popTier)));

    return inv;
}

std::string use_item(Inventory& inv, const std::string& itemId, PlayerCombatSlice& pc) {
    // Find stack
    ItemStack* stack = nullptr;
    for (auto& s : inv.stacks) if (s.id == itemId) { stack = &s; break; }
    if (!stack || stack->count <= 0) return {};

    const ItemDef* def = item_def(itemId);
    if (!def) return {};
    if (def->type != ItemType::Potion && def->type != ItemType::Food) return {};

    std::string msg;
    auto append = [&](const std::string& part) {
        if (!msg.empty()) msg += ", ";
        msg += part;
    };

    if (def->effect.hp != 0) {
        const int healed = std::min(def->effect.hp, pc.maxHp - pc.currentHp);
        pc.currentHp += healed;
        append("+" + std::to_string(healed) + " HP");
    }
    if (def->effect.mp != 0) {
        const int restored = std::min(def->effect.mp, pc.maxMp - pc.currentMp);
        pc.currentMp += restored;
        append("+" + std::to_string(restored) + " MP");
    }
    if (def->effect.sp != 0) {
        const int restored = std::min(def->effect.sp, pc.maxSp - pc.currentSp);
        pc.currentSp += restored;
        append("+" + std::to_string(restored) + " SP");
    }

    inv.remove(itemId, 1);

    std::string head = "Used ";
    head += def->name;
    if (msg.empty()) return head;
    return head + ": " + msg;
}

} // namespace sm
